opm/autodiff/StandardWell_impl.hpp

```diff
@@ -1318,6 +1292,188 @@ namespace Opm
 
 
 
+    template<typename TypeTag>
+    void
+    StandardWell<TypeTag>::
+    checkWellOperability(const Simulator& ebos_simulator)
+    {
+        // TODO: this function is probably can split another function out so that
+        // wellTestingPhysical can share some code with this function
+        // on solution is that this function will be called updateWellOperability
+        // and the actual checking part become another function checkWellOperability
+        // Let us wait until finishing the wellTestingPhysical first.
+
+        // focusing on PRODUCER for now
+        if (well_type_ == INJECTOR) {
+            return;
+        }
+
+        if (!this->underPredictionMode() ) {
+            return;
+        }
+
+        const bool old_well_operable = this->operability_status_.isOperable();
+
+        this->operability_status_.reset();
+
+        updateIPR(ebos_simulator);
+
+        // checking the BHP limit related
+        checkOperabilityUnderBHPLimitProducer(ebos_simulator);
+
+        // TODO: if the BHP limit does not work anyway, we do not need to do the following
+        // We do it now for studying purpose.
+        // checking whether the well can operate under the THP constraints.
+        if (this->wellHasTHPConstraints()) {
+            checkOperabilityUnderTHPLimitProducer(ebos_simulator);
+        }
+
+        // checking whether the well can not produce/inject
+        this->operability_status_.existing_drawdown_correct_direction =
+                                                   ! allDrawDownWrongDirection(ebos_simulator);
+
+        const bool well_operable = this->operability_status_.isOperable();
+
+        if (!well_operable && old_well_operable) {
+            OpmLog::info(" well " + name() + " gets SHUT during iteration ");
+        } else if (well_operable && !old_well_operable) {
+            OpmLog::info(" well " + name() + " gets REVIVED during iteration ");
+        }
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    void
+    StandardWell<TypeTag>::
+    checkOperabilityUnderBHPLimitProducer(const Simulator& ebos_simulator)
+    {
+        const double bhp_limit = mostStrictBhpFromBhpLimits();
+        // TODO: a better way to detect whether the BHP is defaulted or not
+        if ( bhp_limit > 1.5e5 || !this->wellHasTHPConstraints() ) {
```

Comment:
Must use unit system for barsa rather than hard-coded SI constant. To make logic clearer make a bool for the first condition:
```
const bool bhp_limit_defaulted = bhp_limit > 1.5 * unit::barsa; // Crude but works: default is one atmosphere.
```